File: Sample Code/ObjcSampleCode/DJISdkDemo/Demo/Gimbal/GimbalCapability/GimbalCapabilityViewController.m
//
//  GimbalCapabilityViewController.m
//  DJISdkDemo
//
//  Copyright © 2016 DJI. All rights reserved.
//
/**
 *  This file demonstrates how to use gimbal capability to check if a feature is supported by the connected gimbal and the valid range for 
 *  the feature. This demo will do the following steps: 
 *  1. Enable/disable buttons according to the supported features. 
 *  2. For products that support pitch range extension, the program will enable this feature. 
 *  3. When a button is pressed, this demo will get the min or max valid value and rotate the gimbal to the value. 
 *
 *  A feature is represented by a key with DJIGimbalParam prefix. The value in the gimbalCapability dictionary is an istance of
 *  DJIParamCapability or its subclass. A category, capabilityCheck, of DJIGimbal is provided in this demo.
 */
#import <DJISDK/DJISDK.h>
#import "DJIGimbal+CapabilityCheck.h"
#import "GimbalCapabilityViewController.h"
#import "DemoUtility.h"

@interface GimbalCapabilityViewController ()

@property (weak, nonatomic) IBOutlet UIButton *pitchMinButton;
@property (weak, nonatomic) IBOutlet UIButton *pitchMaxButton;
@property (weak, nonatomic) IBOutlet UIButton *yawMinButton;
@property (weak, nonatomic) IBOutlet UIButton *yawMaxButton;
@property (weak, nonatomic) IBOutlet UIButton *rollMinButton;
@property (weak, nonatomic) IBOutlet UIButton *rollMaxButton;

@property (assign, nonatomic) NSNumber *pitchRotation;
@property (assign, nonatomic) NSNumber *yawRotation;
@property (assign, nonatomic) NSNumber *rollRotation;

- (IBAction)rotateGimbalToMin:(id)sender;
- (IBAction)rotateGimbalToMax:(id)sender;
- (IBAction)resetGimbal:(id)sender;

@end

@implementation GimbalCapabilityViewController

-(void) viewWillAppear:(BOOL)animated {
    [super viewWillAppear:animated];
    [self setupButtons];
    [self setupRotationStructs];
    [self enablePitchExtensionIfPossible];
}

-(void) viewWillDisappear:(BOOL)animated {
    [super viewWillDisappear:animated];
}

-(void) setupButtons {
    DJIGimbal* gimbal = [DemoComponentHelper fetchGimbal];
    [self.pitchMinButton setEnabled:[gimbal isFeatureSupported:DJIGimbalParamAdjustPitch]];
    [self.pitchMaxButton setEnabled:[gimbal isFeatureSupported:DJIGimbalParamAdjustPitch]];
    [self.yawMinButton setEnabled:[gimbal isFeatureSupported:DJIGimbalParamAdjustYaw]];
    [self.yawMaxButton setEnabled:[gimbal isFeatureSupported:DJIGimbalParamAdjustYaw]];
    [self.rollMinButton setEnabled:[gimbal isFeatureSupported:DJIGimbalParamAdjustRoll]];
    [self.rollMaxButton setEnabled:[gimbal isFeatureSupported:DJIGimbalParamAdjustRoll]];
}

-(void) setupRotationStructs {
    DJIGimbal* gimbal = [DemoComponentHelper fetchGimbal];
    self.pitchRotation = [gimbal isFeatureSupported:DJIGimbalParamAdjustPitch] ? @(0) : nil;
    self.yawRotation = [gimbal isFeatureSupported:DJIGimbalParamAdjustYaw] ? @(0) : nil;
    self.rollRotation = [gimbal isFeatureSupported:DJIGimbalParamAdjustRoll] ? @(0) : nil;
}

-(void) enablePitchExtensionIfPossible {
    DJIGimbal* gimbal = [DemoComponentHelper fetchGimbal];
    if (gimbal == nil) {
        return;
    }
    BOOL isPossible = [gimbal isFeatureSupported:DJIGimbalParamPitchRangeExtensionEnabled];
    if (isPossible) {
        [gimbal setPitchRangeExtensionEnabled:YES withCompletion:nil];
    }
}


- (IBAction)rotateGimbalToMin:(id)sender {
    DJIGimbal* gimbal = [DemoComponentHelper fetchGimbal];
    if (gimbal == nil) {
        return;
    }
    
    NSString *key = [self getCorrespondingKeyWithButton:(UIButton *)sender];
    NSInteger min = [[gimbal getParamMin:key] integerValue];
    
    if ([key isEqualToString:DJIGimbalParamAdjustPitch]) {
        self.pitchRotation = self.pitchRotation ? @(min) : nil;
    }
    else if ([key isEqualToString:DJIGimbalParamAdjustYaw]) {
        self.yawRotation = self.yawRotation ? @(min) : nil;
    }
    else if ([key isEqualToString:DJIGimbalParamAdjustRoll]) {
        self.rollRotation = self.rollRotation ? @(min) : nil;
    }
    
    [self sendRotateGimbalCommand];
}

- (IBAction)rotateGimbalToMax:(id)sender {
    DJIGimbal* gimbal = [DemoComponentHelper fetchGimbal];
    if (gimbal == nil) {
        return;
    }
    
    NSString *key = [self getCorrespondingKeyWithButton:(UIButton *)sender];
    NSInteger max = [[gimbal getParamMax:key] integerValue];
    
    if ([key isEqualToString:DJIGimbalParamAdjustPitch]) {
        self.pitchRotation = self.pitchRotation ? @(max) : nil;
    }
    else if ([key isEqualToString:DJIGimbalParamAdjustYaw]) {
        self.yawRotation = self.yawRotation ? @(max) : nil;
    }
    else if ([key isEqualToString:DJIGimbalParamAdjustRoll]) {
        self.rollRotation = self.rollRotation ? @(max) : nil;
    }
    
    [self sendRotateGimbalCommand];
}

- (IBAction)resetGimbal:(id)sender {
    self.pitchRotation = self.pitchRotation ? @(0) : nil;
    self.yawRotation = self.yawRotation ? @(0) : nil;
    self.rollRotation = self.rollRotation ? @(0) : nil;

    [self sendRotateGimbalCommand];
}

-(void) sendRotateGimbalCommand {
    DJIGimbal* gimbal = [DemoComponentHelper fetchGimbal];
    if (gimbal == nil) {
        return;
    }

    DJIGimbalRotation *rotation = [DJIGimbalRotation gimbalRotationWithPitchValue:self.pitchRotation
                                                                        rollValue:self.rollRotation
                                                                         yawValue:self.yawRotation time:2
                                                                             mode:DJIGimbalRotationModeAbsoluteAngle];
    [gimbal rotateWithRotation:rotation completion:^(NSError * _Nullable error) {
        if (error) {
            ShowResult(@"rotateWithRotation failed: %@", error.description);
        }
    }];
}

-(NSString *) getCorrespondingKeyWithButton:(UIButton *)button {
    if (button == self.pitchMinButton || button == self.pitchMaxButton) {
        return DJIGimbalParamAdjustPitch;
    }
    else if (button == self.yawMinButton || button == self.yawMaxButton) {
        return DJIGimbalParamAdjustYaw;
    }
    else if (button == self.rollMinButton || button == self.rollMaxButton) {
        return DJIGimbalParamAdjustRoll;
    }
    return nil;
}

@end

File: Sample Code/SwiftSampleCode/DJISDKSwiftDemo/Components/Gimbal/GimbalCapability/GimbalCapabilityViewController.swift
//
//  GimbalCapabilityViewController.swift
//  DJISDKSwiftDemo
//
//  Copyright © 2016 DJI. All rights reserved.
//
/**
*  This file demonstrates how to use gimbal capability to check if a feature is supported by the connected gimbal and the valid range for
*  the feature. This demo will do the following steps:
*  1. Enable/disable buttons according to the supported features.
*  2. For products that support pitch range extension, the program will enable this feature.
*  3. When a button is pressed, this demo will get the min or max valid value and rotate the gimbal to the value.
*
*  A feature is represented by a key with DJIGimbalParam prefix. The value in the gimbalCapability dictionary is an istance of
*  DJIParamCapability or its subclass. A category, capabilityCheck, of DJIGimbal is provided in this demo.
*/

import DJISDK

class GimbalCapabilityViewController: DJIBaseViewController{
    @IBOutlet weak var pitchMinButton:UIButton!
    @IBOutlet weak var pitchMaxButton:UIButton!
    @IBOutlet weak var yawMinButton:UIButton!
    @IBOutlet weak var yawMaxButton:UIButton!
    @IBOutlet weak var rollMinButton:UIButton!
    @IBOutlet weak var rollMaxButton:UIButton!
    
    var pitchRotation : DJIGimbalAngleRotation! = DJIGimbalAngleRotation()
    var yawRotation : DJIGimbalAngleRotation! = DJIGimbalAngleRotation()
    var rollRotation : DJIGimbalAngleRotation! = DJIGimbalAngleRotation()
    
    
    override func viewWillAppear(_ animated: Bool) {
        super.viewWillAppear(animated)
        self.setupButtons()
        self.setupRotationStructs()
        self.enablePitchExtensionIfPossible()
    }
    
    override func viewWillDisappear(_ animated: Bool) {
        super.viewWillDisappear(animated)
    }
    
    func setupButtons() {
        let gimbal : DJIGimbal? = self.fetchGimbal()
        if (gimbal == nil) {
            self.pitchMinButton.isEnabled = false
            self.pitchMaxButton.isEnabled = false
            self.yawMinButton.isEnabled = false
            self.yawMaxButton.isEnabled = false
            self.rollMinButton.isEnabled = false
            self.rollMaxButton.isEnabled = false
        }else {
            self.pitchMinButton.isEnabled = gimbal!.isFeatureSupported(DJIGimbalParamAdjustPitch)
            self.pitchMaxButton.isEnabled = gimbal!.isFeatureSupported(DJIGimbalParamAdjustPitch)
            self.yawMinButton.isEnabled = gimbal!.isFeatureSupported(DJIGimbalParamAdjustYaw)
            self.yawMaxButton.isEnabled = gimbal!.isFeatureSupported(DJIGimbalParamAdjustYaw)
            self.rollMinButton.isEnabled = gimbal!.isFeatureSupported(DJIGimbalParamAdjustRoll)
            self.rollMaxButton.isEnabled = gimbal!.isFeatureSupported(DJIGimbalParamAdjustRoll)
        }
    }
    
    func setupRotationStructs() {
        let gimbal: DJIGimbal? = self.fetchGimbal()
        self.pitchRotation.enabled = ObjCBool(false)
        self.yawRotation.enabled = ObjCBool(false)
        self.rollRotation.enabled = ObjCBool(false)
        
        if (gimbal != nil) {
            self.pitchRotation.enabled = ObjCBool(gimbal!.isFeatureSupported(DJIGimbalParamAdjustPitch))
            self.yawRotation.enabled = ObjCBool(gimbal!.isFeatureSupported(DJIGimbalParamAdjustYaw))
            self.rollRotation.enabled = ObjCBool(gimbal!.isFeatureSupported(DJIGimbalParamAdjustRoll))
        }
    }
    
    func enablePitchExtensionIfPossible() {
        let gimbal: DJIGimbal? = self.fetchGimbal()
        if (gimbal == nil) {
            return
        }
        let isPossible: Bool = gimbal!.isFeatureSupported(DJIGimbalParamPitchRangeExtensionEnabled)
        if isPossible {
            gimbal!.setPitchRangeExtensionEnabled(true, withCompletion: nil)
        }
    }
    
    @IBAction func rotateGimbalToMin(_ sender: AnyObject) {
        let gimbal: DJIGimbal? = self.fetchGimbal()
        if gimbal == nil {
            return
        }
        let key: String = self.getCorrespondingKeyWithButton((sender as! UIButton))!
        let min: Int? = gimbal!.getParamMin(key)
        if (key == DJIGimbalParamAdjustPitch && min != nil) {
            self.pitchRotation.direction = DJIGimbalRotateDirection.clockwise
            self.pitchRotation.angle = Float(min!)
        }
        else if (key == DJIGimbalParamAdjustYaw && min != nil) {
            self.yawRotation.direction = DJIGimbalRotateDirection.clockwise
            self.yawRotation.angle = Float(min!)
        }
        else if (key == DJIGimbalParamAdjustRoll && min != nil) {
            self.rollRotation.direction = DJIGimbalRotateDirection.clockwise
            self.rollRotation.angle = Float(min!)
        }
        
        self.sendRotateGimbalCommand()
    }
    
    @IBAction func rotateGimbalToMax(_ sender: AnyObject) {
        let gimbal: DJIGimbal? = self.fetchGimbal()
        if gimbal == nil {
            return
        }
        let key: String = self.getCorrespondingKeyWithButton((sender as! UIButton))!
        let max: Int? = gimbal!.getParamMax(key)
        if (key == DJIGimbalParamAdjustPitch && max != nil) {
            self.pitchRotation.direction = DJIGimbalRotateDirection.clockwise
            self.pitchRotation.angle = Float(max!)
        }
        else if (key == DJIGimbalParamAdjustYaw && max != nil) {
            self.yawRotation.direction = DJIGimbalRotateDirection.clockwise
            self.yawRotation.angle = Float(max!)
        }
        else if (key == DJIGimbalParamAdjustRoll && max != nil) {
            self.rollRotation.direction = DJIGimbalRotateDirection.clockwise
            self.rollRotation.angle = Float(max!)
        }
        
        self.sendRotateGimbalCommand()
    }
    
    @IBAction func resetGimbal(_ sender: AnyObject) {
        self.pitchRotation.angle = 0
        self.yawRotation.angle = 0
        self.rollRotation.angle = 0
        self.sendRotateGimbalCommand()
    }
    
    func sendRotateGimbalCommand() {
        let gimbal: DJIGimbal? = self.fetchGimbal()
        if gimbal == nil {
            return
        }
        gimbal!.rotateGimbal(with: DJIGimbalRotateAngleMode.angleModeAbsoluteAngle, pitch: self.pitchRotation, roll: self.rollRotation, yaw: self.yawRotation, withCompletion: {(error: Error?) -> Void in
            if error != nil {
                self.showAlertResult("rotateGimbalWithAngleMode failed: \(error!)")
            }
        })
    }
    
    func getCorrespondingKeyWithButton(_ button: UIButton) -> String? {
        if button == self.pitchMinButton || button == self.pitchMaxButton {
            return DJIGimbalParamAdjustPitch
        }
        else if button == self.yawMinButton || button == self.yawMaxButton {
            return DJIGimbalParamAdjustYaw
        }
        else if button == self.rollMinButton || button == self.rollMaxButton {
            return DJIGimbalParamAdjustRoll
        }
        
        return ""
    }
}
